aptos-move/mvhashmap/src/versioned_module_storage.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::{Flag, ShiftedTxnIndex, TxnIndex};
+use claims::assert_some;
+use crossbeam::utils::CachePadded;
+use dashmap::DashMap;
+use std::{collections::BTreeMap, hash::Hash, ops::Deref, sync::Arc};
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ModuleReadError {
+    /// There is no entry for this module yet, and it has to be initialized.
+    Uninitialized,
+    /// A dependency on some other transaction has been found, or there is a
+    /// pending code publish which has not yet been committed.
+    Dependency(TxnIndex),
+}
+
+/// An entry in versioned module storage. We distinguish between modules (and
+/// whatever other associated information that is cached) which are:
+///   1. Pending to be published, i.e., writes created but the transaction has not
+///      yet been committed. Reading a pending module always yields a dependency.
+///   2. Published - a module that has been committed, and we are guaranteed there
+///      is no speculative information.
+enum Entry<M> {
+    Pending { flag: Flag, entry: Arc<M> },
+    Published { entry: Arc<M> },
+}
+
+impl<M: Hash> Entry<M> {
+    fn mark_estimate(&mut self) {
```

Comment:
This is confusing because we are re-using estimate flag for a different purpose here: originally value written at the end of the execution is considered Default = Done, while abort marks them as estimates. Observing the estimate then means a dependency, and they are cleared by the next re-execution.

Here, pending already means an estimate, i.e. that whoever sees pending returns a Dependency, so the only functionality that is missing from the caller point of view, assuming the dependencies should outlive speculation re-executions, is to publish a subset of pending modules, and remove others (during commit) - in that case it might even make sense to record the list of all pending modules for a transaction here (to live across speculative boundaries). We could also choose to clear pending modules alongside other estimates (e.g. from data MVHashMap) after each re-execution, which seems to be the intent here (given the remove API). Worth thinking about. But for this comment's purposes, in any case, marking an Estimate only serves as a way to check an invariant (that whatever we published was from the latest incarnation) - and in that case Done and Estimate are no longer good Flag names.